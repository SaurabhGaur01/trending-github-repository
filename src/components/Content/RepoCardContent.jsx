import React, { useState, useEffect } from "react";
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import '../RepoApp.scss';
import GitHubIcon from '@material-ui/icons/GitHub';
import Rating from '@material-ui/lab/Rating';
import { setStarredData, clearStarredData } from '../../ducks/starredData';

const RepoCardContent = ({ 
    repository, 
    actionSetStarredData, 
    starredData,
    actionClearStarredData 
}) =>{
    const {
        id,
        name,
        description,
        html_url,
        stargazers_count,
      } = repository;
    const [starred, setStarred] = useState({});
    useEffect(() => {
        let isStarred = starredData.find((repo) => repo.id === id);
        isStarred ? setStarred(1) : setStarred(null);
    }, [starredData, id]);

    const handleClick = (event, newValue, id) => {
        setStarred(newValue);
        if (newValue) {
            const selectedStarredData = [...starredData, repository];
            actionSetStarredData(selectedStarredData);
        }else {
            const selectedStarredData = starredData.filter(repo => repo.id !== id);
            actionClearStarredData();
            actionSetStarredData(selectedStarredData);
        }
    }
    
    return (
        <Card className="card-container" key={id}>
            <CardContent>
                <Rating
                    name={`rating-card-${id}`}
                    max={1}
                    onChange={(event, newValue) => handleClick(event, newValue, id)}
                    value={starred ? 1: 0}
                />

                <div className="first-section">
                    <div className="heading-label">{name}</div>
                    <div className="description">{description}</div>
                </div>
                
                <div className="star-count">
                    <hr/>
                        {stargazers_count} stars
                    <hr/> 
                </div>

                <div className="url">
                    <a href={html_url} target="_blank" rel="noopener noreferrer">
                        <GitHubIcon/>Github Repo
                    </a>
                </div>         
            </CardContent>   
        </Card>
    )
};

RepoCardContent.propTypes = {
    actionSetStarredData: PropTypes.func.isRequired,
    actionClearStarredData: PropTypes.func.isRequired,
    starredData:PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        description: PropTypes.string,
        html_url: PropTypes.string,
        stargazers_count: PropTypes.number,
    })).isRequired,
    repository: PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        description: PropTypes.string,
        html_url: PropTypes.string,
        stargazers_count: PropTypes.number,
    }).isRequired,
};

const mapDispatchAsProps = dispatch => {
    return {
        actionSetStarredData: (value) => dispatch(setStarredData(value)),
        actionClearStarredData: () => dispatch(clearStarredData()),
    };
}

const mapStateToProps = state => ({
    starredData: state.starredData,
});

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

export { RepoCardContent as TestableRepoCardContent };
  
export default hocChain(RepoCardContent);