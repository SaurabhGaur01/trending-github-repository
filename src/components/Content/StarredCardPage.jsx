import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import GitHubIcon from '@material-ui/icons/GitHub';
import Rating from '@material-ui/lab/Rating';
import '../RepoApp.scss';

const StarredCardPage = ({ starredData }) => (
    <div className="main-card">
        { starredData.length > 0 && starredData.map(
            ({
                id,
                name,
                description,
                html_url,
                stargazers_count,
            })=>{   
                return (
                    <Card className="card-container" key={id}>
                        <CardContent>
                            <Rating
                                name={`rating-card-${id}`}
                                max={1}
                                value={1}
                            />
                
                            <div className="first-section">
                                <div className="heading-label">{name}</div>
                                <div className="description">{description}</div>
                            </div>
                        
                            <div className="star-count">
                                <hr/>
                                    {stargazers_count} stars
                                <hr/> 
                            </div>
        
                            <div className="url">
                                <a href={html_url} target="_blank" rel="noopener noreferrer">
                                    <GitHubIcon/>Github Repo
                                </a>
                            </div>         
                        </CardContent>   
                    </Card>               
                )
            }    
        )}
    </div>
);

StarredCardPage.propTypes = {
    starredData:PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        description: PropTypes.string,
        html_url: PropTypes.string,
        stargazers_count: PropTypes.number,
    })).isRequired,
};

const mapStateToProps = state => ({
    starredData: state.starredData,
});

const hocChain = compose(
    connect(mapStateToProps),
);

export { StarredCardPage as TestableStarredCardPage };
  
export default hocChain(StarredCardPage);