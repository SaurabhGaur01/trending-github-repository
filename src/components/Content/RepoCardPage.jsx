import React from "react";
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import RingLoader from "react-spinners/RingLoader";
import { loaderClass, loaderSize, loaderColor } from '../../constants/loaderClass';
import RepoCardContent from "./RepoCardContent";
import { API_CALL_STATUS_LOADING, API_CALL_STATUS_SUCCESS } from '../../constants/status';

const RepoCardPage = ({ repoData, apiCallStatus }) => (
    <div className="main-card">
        <RingLoader
            css={loaderClass}
            size={loaderSize}
            color={loaderColor}
            loading={apiCallStatus === API_CALL_STATUS_LOADING ? true: false}
        />  
        {
            apiCallStatus === API_CALL_STATUS_SUCCESS && repoData.map((repository) => {
                return <RepoCardContent repository={repository} key={repository.id} />;
            })
        }
    </div>    
);

RepoCardPage.propTypes = {
    apiCallStatus: PropTypes.string.isRequired,
    repoData: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        name: PropTypes.string,
        description: PropTypes.string,
        html_url: PropTypes.string,
        stargazers_count: PropTypes.number,
    })).isRequired,
};

export const mapStateToProps = state => ({
    repoData: state.repoData,
    apiCallStatus: state.apiCallStatus,
});

const hocChain = compose(
    connect(mapStateToProps),
);

export { RepoCardPage as TestableRepoCardPage };
  
export default hocChain(RepoCardPage);