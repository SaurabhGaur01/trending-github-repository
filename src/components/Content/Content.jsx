import React from 'react';
import TabPanel from '../Tabs/TabPanel';
import TabHeader from '../Tabs/TabHeader';
import RepoCardPage from './RepoCardPage';
import StarredCardPage from './StarredCardPage';
import '../RepoApp.scss';

const Content = () => {
  const [tabValue, setTabValue] = React.useState(0);
  return (
    <React.Fragment>
      <TabHeader value={tabValue} handleChange={(event, newValue)=>{setTabValue(newValue)}} />
      <TabPanel value={tabValue} index={0}>
          <RepoCardPage />  
      </TabPanel>
      <TabPanel value={tabValue} index={1}>
          <StarredCardPage />
      </TabPanel>
    </React.Fragment>
  );
}

export default Content;