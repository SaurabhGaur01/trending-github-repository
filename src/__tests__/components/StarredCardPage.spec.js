import React from 'react';
import { shallow } from 'enzyme';
import { TestableStarredCardPage, mapStateToProps } from '../../components/Content/StarredCardPage';

describe('StarredCardPage Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            starredData: [
                {
                    id: 1,
                    name: 'mockName',
                    description: 'mockDescription',
                    html_url: 'mockUrl',
                    stargazers_count: 20,
                }
            ],
        }
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableStarredCardPage {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should render correctly when starredData is empty', () => {
        props.starredData = [];
        const renderedModule = shallow(<TestableStarredCardPage {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    describe('mapStateToProps', () => {
        const starredDataMock = 'starredDataMock';

        const mockedState = {
            starredData: starredDataMock,
        };

        it('should return the module state from the redux store', () => {
            const mockPropsFromRedux = mapStateToProps(mockedState);
            expect(mockPropsFromRedux).toEqual(
                {
                    starredData: starredDataMock,
                }
            )
        })
    })
})