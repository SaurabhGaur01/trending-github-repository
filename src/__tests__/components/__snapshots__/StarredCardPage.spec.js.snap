// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarredCardPage Component should render correctly 1`] = `
<div
  className="main-card"
>
  <WithStyles(ForwardRef(Card))
    className="card-container"
    key="1"
  >
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Rating))
        max={1}
        name="rating-card-1"
        value={1}
      />
      <div
        className="first-section"
      >
        <div
          className="heading-label"
        >
          mockName
        </div>
        <div
          className="description"
        >
          mockDescription
        </div>
      </div>
      <div
        className="star-count"
      >
        <hr />
        20
         stars
        <hr />
      </div>
      <div
        className="url"
      >
        <a
          href="mockUrl"
          rel="noopener noreferrer"
          target="_blank"
        >
          <GitHubIcon />
          Github Repo
        </a>
      </div>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(Card))>
</div>
`;

exports[`StarredCardPage Component should render correctly when starredData is empty 1`] = `
<div
  className="empty-msg"
>
  *No data found
</div>
`;
